packages/connections/src/XHRConnection.js

```diff
@@ -0,0 +1,170 @@
+   * @param {string} [options.responseType]
+   */
+  constructor(url, options = {}) {
+    super(url);
+
+    const {
+      method = "GET",
+      body = null,
+      headers = {},
+      responseType = "json"
+    } = options;
+
+    if (ALLOWED_METHODS.includes(method) === -1) {
+      throw new Error("Invalid method for XHR Request.");
+    }
+
+    /**
+     * @property {string}
+     * @protected
+     * @name XHRConnection#method
+     */
+    Object.defineProperty(this, "method", { value: method });
+
+    /**
+     * @property {*}
+     * @protected
+     * @name XHRConnection#body
+     */
+    Object.defineProperty(this, "body", { value: body });
+
+    /**
+     * @property {object}
+     * @protected
+     * @name XHRConnection#headers
+     */
+    Object.defineProperty(this, "headers", { value: headers });
+
+    /**
+     * @property {string}
+     * @protected
+     * @name XHRConnection#responseType
+     */
+    Object.defineProperty(this, "responseType", { value: responseType });
+
+    /**
+     * @property {XMLHttpRequest}                 ̵
+     * @protected
+     * @name XHRConnection#xhr
+     */
+    Object.defineProperty(this, "xhr", { value: new XMLHttpRequest() });
+  }
+
+  get response() {
+    return this.xhr.response;
+  }
+
+  get status() {
+    return this.xhr.status;
+  }
+
+  /**
+   * create, prepare, open and send the xhr request
+   * @param {string} [token] – authentication token
+   */
+  open(token) {
+    super.open(token);
+
+    this.xhr.addEventListener("progress", () => {
+      this.emit(
+        ConnectionEvent.DATA,
+        new ConnectionEvent(this, ConnectionEvent.DATA)
+      );
+    });
+
+    this.xhr.addEventListener("load", () => {
+      this.state = XHRConnection.CLOSED;
+
+      if (this.status >= 400) {
+        this.emit(
+          ConnectionEvent.ERROR,
+          new ConnectionEvent(this, ConnectionEvent.ERROR)
+        );
+
+        return;
+      }
+
+      this.emit(
+        ConnectionEvent.COMPLETE,
+        new ConnectionEvent(this, ConnectionEvent.COMPLETE)
+      );
+    });
+
+    this.xhr.addEventListener("abort", () => {
+      this.state = XHRConnection.CLOSED;
+
+      this.emit(
+        ConnectionEvent.ABORT,
+        new ConnectionEvent(this, ConnectionEvent.ABORT)
+      );
+    });
+
+    this.xhr.addEventListener("error", () => {
+      this.state = XHRConnection.CLOSED;
+
+      this.emit(
+        ConnectionEvent.ERROR,
+        new ConnectionEvent(this, ConnectionEvent.ERROR)
+      );
+    });
+
+    this.xhr.addEventListener("timeout", () => {
+      this.state = XHRConnection.CLOSED;
+
+      this.emit(
+        ConnectionEvent.ERROR,
+        new ConnectionEvent(this, ConnectionEvent.ERROR)
+      );
+    });
+
+    this.state = XHRConnection.OPEN;
+    this.xhr.open(this.method, this.url);
+    this.emit(
+      ConnectionEvent.OPEN,
+      new ConnectionEvent(this, ConnectionEvent.OPEN)
+    );
+
+    Object.keys(this.headers).forEach(key => {
```

Comment:
We should check `headers` is an Object before we set it in the constructor.

Reply:
Yep, will add a check.

Reply:
✔️  Done